[FIX] point_of_sale: prevent error when reopening a POS session

adip-odoo opened the pull request:
Currently, an error occurs on reopening a POS session if its 'Login Number Sequence' record has been deleted. This results in preventing the user from accessing the session.

**Steps to produce:**
- Install the `point_of_sale` module (with demo data).
- Open the register of the **Furniture Shop** from the POS dashboard.
- Now, without closing the register of shop, Navigate to:
 `Settings > Technical > Sequences & Identifiers > Sequences (list)`.
- Delete the `Login Number Sequence of Session 3` sequence record.
- Reopen that shop session again.
- Observe the error at the backend.

**Error:**
```
UndefinedFunction
operator does not exist: integer = boolean
LINE 1: SELECT number_next FROM ir_sequence WHERE id=false FOR UPDAT...
                                                    ^
HINT:  No operator matches the given name and argument types. You might need
to add explicit type casts.
```

The error occurs because the system attempts to access the `login_number_seq_id` of the POS session at [1], but it is unavailable as the user has already deleted it.

This commit adds validation to prevent deletion of the 'Login Number Sequence' if the associated POS session is not closed.

[1] - https://github.com/odoo/odoo/blob/948f6c5afdf76b9a25daa120fec1edd21c5c08d6/addons/point_of_sale/models/pos_session.py#L404-L406

Sentry-6276299189


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/197715.png)](https://mergebot.odoo.com/odoo/odoo/pull/197715)

adip-odoo commented:
Hello @odoo/sentry_reviewers

Please review this PR.

Thanks!

caburj commented:
Hi @adip-odoo. Can you also add a check before unlinking a sequence record? We should prevent deletion of a sequence when it's linked to an open session.

adip-odoo commented:
Hello @caburj,

I have updated the commit as you suggested and added the check to prevent the deletion of a sequence when it's linked to an open session.

Please, review it.

Thanks!

maan-odoo commented on addons/point_of_sale/models/ir_sequence.py:
```diff
@@ -0,0 +1,13 @@
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_sequence(self):
+        for seq in self:
+            pos = self.env['pos.session'].search(domain=[('login_number_seq_id', '=', seq.id)], limit=1)
+            if not self._context.get('delete_from_unlink_session') and pos and pos.state != 'closed':
+                raise UserError(_('You cannot delete a sequence that is currently in use by a POS session.'))
```
```suggestion
        sessions = self.env['pos.session'].search(domain=[('login_number_seq_id', 'in', self.ids)])
        acive_sessions = sessions.filtered(lambda s: s.state != 'closed')
        if acive_sessions and not self.env.context.get('force_unlink'):
            raise UserError(_(
                "You cannot delete a sequence used in an active session: %s",
                acive_sessions.login_number_seq_id.mapped('name')
            ))
```
- avoid search in loop

caburj commented on addons/point_of_sale/models/pos_session.py:
```diff
@@ -398,7 +398,7 @@ def create(self, vals_list):
     def unlink(self):
         self.statement_line_ids.unlink()
-        (self.order_seq_id | self.login_number_seq_id).unlink()
+        (self.order_seq_id | self.login_number_seq_id).with_context(force_unlink=True).unlink()
         return super(PosSession, self).unlink()
```
```suggestion
        result = super(PosSession, self).unlink()
        (self.order_seq_id | self.login_number_seq_id).unlink()
        return result
```

I think you need to switch the deletion so that no need to `force_unlink`.

caburj commented on addons/point_of_sale/models/ir_sequence.py:
```diff
@@ -0,0 +1,16 @@
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_sequence(self):
+        sessions = self.env['pos.session'].search(domain=[('login_number_seq_id', 'in', self.ids)])
```
I just noticed that you forgot to include in the check the `order_seq_id`.

adip-odoo commented:
Hello @caburj,

I have applied the changes. Can you please review it?

Thanks!

adgu-odoo reviewed (approved):
robodoo r+
